Add hybrid rendering docs

alan-agius4 opened the pull request:
**build(docs-infra): include API references for `@angular/ssr` entry points**
This commit utilizes the `@angular/ssr` NPM package to generate an API reference for its entry points using the `generate_api_docs` Bazel rule, which will be included in http://angular.dev/api.

**docs: add hybrid rendering API guide**
This commit introduces a new guide for hybrid rendering APIs, which are currently in developer preview. This documentation provides insights into the usage and features of the APIs, helping developers understand their capabilities and limitations during the preview phase.

**docs: update SSR documentation to include API references**
This update removes hard-coded options and enhances the SSR documentation to incorporate API references.


**refactor(compiler-cli): Improved `filterMethodOverloads` to Include Members without body**
Previously, `filterMethodOverloads` excluded all members without a body, causing issues with the extraction of functions and members in TypeScript types.

alan-agius4 committed:
build(docs-infra): include API references for `@angular/ssr` entry points

This commit utilizes the `@angular/ssr` NPM package to generate an API reference for its entry points using the `generate_api_docs` Bazel rule, which will be included in http://angular.dev/api.

alan-agius4 committed:
docs: update SSR documentation to include API references

This update removes hard-coded options and enhances the SSR documentation to incorporate API references.

alan-agius4 committed:
docs: add hybrid rendering API guide

This commit introduces a new guide for hybrid rendering APIs, which are currently in developer preview. This documentation provides insights into the usage and features of the APIs, helping developers understand their capabilities and limitations during the preview phase.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,231 @@
+
+export const serverRoutes: ServerRoute[] = [
+  {
+    path: 'blog/:id',
```
Nit: This isn't super important but `/blog/2` implies a site with multiple blogs which is just a little weird to conceptualize. Maybe `/posts/2` or `/articles/2` would be easier for users to imagine?

dgp1130 reviewed (commented):
Thanks for putting all this together! I agree a separate page is helpful given the state of this API. Lots of great details here to guide users.

Once we stabilize this API, we should consider following some of these into our existing SSR pages and thinking about how we can section and segment this information in a reasonable way (ex. SSR routing, working with requests/responses, integrating your own server, etc.)

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,231 @@
+    path: '**',
+    renderMode: RenderMode.Server,
+  },
+];
```
Yes we changed that some time ago, to be closer to the Angular router routes.

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,231 @@
+/**
+ * This is a request handler used by the Angular CLI (dev-server and during build).
+ */
+export const reqHandler = createRequestHandler(async (req) => {
```
Yeah, `reqHandler` needs to be exported for the dev-server.

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
````diff
@@ -0,0 +1,231 @@
+};
+```
+
+## Configuring Server Routes
````
Agreed

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,231 @@
+const serverConfig: ApplicationConfig = {
+  providers: [
+    provideServerRendering(),
+    provideServerRoutesConfig(serverRoutes),
```
No, but we should!

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
````diff
@@ -0,0 +1,231 @@
+];
+```
+
+## Entry-points
````
Yeah potentially we do split it out.

devversion commented on packages/BUILD.bazel:
```diff
@@ -106,3 +109,34 @@ generate_api_manifest(
 )
+
+### @angular/ssr package located in the Angular/CLI repository
+generate_api_docs(
```
Should we have a folder like `packages/ssr/BUILD` where we store this better?

devversion reviewed (approved):
LGTM For Bazel

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
````diff
@@ -0,0 +1,231 @@
+const reqHandler = createNodeRequestHandler(app);
+```
+
+### @angular/ssr/tokens
````
Good point

alan-agius4 committed:
refactor(compiler-cli): Improved `filterMethodOverloads` to Include Members without body

Previously, `filterMethodOverloads` excluded all members without a body, causing issues with the extraction of functions and members in TypeScript types.

alan-agius4 commented:
@devversion, can you also PTAL at the compiler changes? Thanks

dgp1130 reviewed (approved):
Thanks for updating this, I left a handful of suggestions but nothing too major. I really like the described trade-offs between the different render modes, I think that's going to be really helpful for devs to understand which mode is right for their particular use case.

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,301 @@
+
+**IMPORTANT**: When employing `inject` inside `getPrenderParams`, please remember that `inject` must be used synchronously. It cannot be invoked within asynchronous callbacks or following any `await` statements. For more information, refer to [`runInInjectionContext` API reference](api/core/runInInjectionContext).
+
+**NOTE**: [`getPrerenderParams`](api/ssr/ServerRoutePrerenderWithParams#getPrerenderParams 'API reference') can only be used when `renderMode` is set to [`RenderMode.Prerender`](api/ssr/RenderMode#Prerender 'API reference'). This function runs during the build process, so it should not rely on any browser-specific APIs or data. If the [`fallback`](api/ssr/ServerRoutePrerenderWithParams#fallback ' ...
```
We rely on type safety for this.

JeanMeche commented on adev/src/content/guide/ssr.md:
````diff
@@ -75,26 +59,23 @@ Hydration is the process that restores the server side rendered application on t
 
-<docs-code language="typescript">
-
+```typescript
````
fwiw, the shiki the highlighter, supports `angular-ts` a language, which does highlight the template correctly.

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts:
```diff
@@ -271,26 +271,21 @@ class ClassExtractor {
+          ts.isMethodDeclaration(nextDeclaration) &&
+          nextDeclaration.name.getText() === declaration.name?.getText();
+
+        // Return only the first occurrence of an abstract method to avoid overload duplication.
```
wouldn't this be the "last" occurence?

alan-agius4 replied:
Yeah :)

devversion reviewed (approved):
LGTM for compiler + bazel; but it's interesting to note that API extraction is now run against `.d.ts`. Not sure what implications this would have; but we can just try I guess

alan-agius4 commented:
> Not sure what implications this would have; but we can just try I guess

@devversion, I am not sure either, but at least for the @angular/ssr package it seems to work pretty good.

devversion commented:
I think we are fine trying if it works 😄 — but it's maybe something we should document in the api-extractor folder?

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,301 @@
+  - **Ideal for:** Providing a fast initial load experience.
+  - **Benefits:**
+      - Users see a basic layout quickly, improving perceived performance.
+  - **Learn more**: [App Shell guide](ecosystem/service-workers/app-shell)
```
Added

alan-agius4 commented:
@dgp1130, can you please take a look at the last commit?

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -35,52 +35,23 @@ To add SSR to an **existing** project, use the Angular CLI `ng add` command.
 
 The Server Routing Configuration allows you to specify how each route in your application should be rendered. The following rendering modes are available:
+- **Server (SSR):** Renders your application on the server for each request, sending a fully populated HTML page to the browser. This is beneficial for SEO and initial load time. See the [Server-Side Rendering (SSR) guide](guide/ssr) for more information.
+- **Client (CSR):** Renders your application in the browser (this is the default Angular behavior). This is suitable for highly interactive applications.
```
Suggestion: I'd still avoid "highly interactive" as a reason for preferring CSR, given that SSR support doesn't really restrict the interactivity of an application. Should we briefly mention data locality and user interactions here?

jelbourn replied:
+1, my suggestion here is 
```suggestion
- **Client (CSR):** Renders your application in the browser. This is the default Angular behavior.
```

* goes along with my comment above on removing any recommendations from this section
* Parenthesis in our prose content should be _relatively_ rare, only used when a regular sentence or subclause is insufficient

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -35,52 +35,23 @@ To add SSR to an **existing** project, use the Angular CLI `ng add` command.
+- **Server (SSR):** Renders your application on the server for each request, sending a fully populated HTML page to the browser. This is beneficial for SEO and initial load time. See the [Server-Side Rendering (SSR) guide](guide/ssr) for more information.
+- **Client (CSR):** Renders your application in the browser (this is the default Angular behavior). This is suitable for highly interactive applications.
+- **Prerender (SSG):** Prerenders your application at build time, generating static HTML files for each route. This offers the fastest loading times and is ideal for static content. See the [Prerendering guide](guide/prerendering) for more information.
+- **App Shell:** Generates a static HTML shell of your application during the build process. This shell, a basic layout of your app (header, navigation, footer, etc.), is displayed immediately for CSR routes while the full application loads in the background. This improves perceived performance by giving the user something to look at right away. See the [App Shell guide](ecosystem/service-workers ...
```
Consider: "This shell... is displayed immediately for CSR routes..." implies that `RenderMode.Client` uses an app shell. I think dropping "for CSR routes" here might help keep users in the mind of `RenderMode.AppShell`.

alan-agius4 replied:
When app-shell is enabled, it will be applied to all CSR routes.

dgp1130 replied:
I thought `renderMode: RenderMode.AppShell` applied only for the route it matches? Does it do more than that?

alan-agius4 replied:
An application can only have a one AppShell, which is applied across all CSR (Client-Side Rendering) routes. By setting `renderMode` to `AppShell`, you configuring that route as the application's shell, to be used by all CSR routes.

```ts
{
  path: 'shell',
  renderMode: RenderMode.AppShell,
}, {
  path: 'admin/**',
  renderMode: RenderMode.Client,
}
```

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -35,52 +35,23 @@ To add SSR to an **existing** project, use the Angular CLI `ng add` command.
+- **Server (SSR):** Renders your application on the server for each request, sending a fully populated HTML page to the browser. This is beneficial for SEO and initial load time. See the [Server-Side Rendering (SSR) guide](guide/ssr) for more information.
+- **Client (CSR):** Renders your application in the browser (this is the default Angular behavior). This is suitable for highly interactive applications.
+- **Prerender (SSG):** Prerenders your application at build time, generating static HTML files for each route. This offers the fastest loading times and is ideal for static content. See the [Prerendering guide](guide/prerendering) for more information.
+- **App Shell:** Generates a static HTML shell of your application during the build process. This shell, a basic layout of your app (header, navigation, footer, etc.), is displayed immediately for CSR routes while the full application loads in the background. This improves perceived performance by giving the user something to look at right away. See the [App Shell guide](ecosystem/service-workers ...
```
Consider: "This improves perceived performance by giving the user something to look at right away." also applies to SSR, so a quick distinction might be helpful here. Maybe adding that this is able to render the page scaffolding without needing to do a full page render or have access to all required data would help differentiate it a bit?

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+### Choosing the right rendering mode
+
+Now that you understand the different rendering modes, let's explore how to choose the best one for your application. The optimal choice depends on the specific needs and priorities of your project. Here's a breakdown of the key considerations:
+Each rendering mode has its own strengths and weaknesses. Consider these factors when making your decision:
```
Suggestion: I think this line will be folded into the preceding paragraph due to the lack of a newline. Alternatively we could just drop this line entirely? Not sure how necessary it is.

alan-agius4 replied:
Removed the entire line.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+Each rendering mode has its own strengths and weaknesses. Consider these factors when making your decision:
+
+* **Performance:** How quickly do you need the page to load for users?
+* **SEO:** How important is search engine visibility for this content?
```
Suggestion: Spell out "Search Engine Optimization (SEO)" either here or earlier in the doc for those who might not be familiar with the acronym.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+
+By evaluating these factors, you can determine the optimal rendering strategy for each route of your application.
+
+##### Server (SSR)
```
Nit: I think these can be h4 rather than h5, unless you were specifically going for h5 to keep them out of the navigation pane or something?

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+##### Client (CSR)
+- **Ideal for:**
+  - Routes where data primarily resides on the client-side and is not known to the server (e.g., PWAs with offline functionality)
+  - Routes with deep user interactions where constant server updates provide little benefit (e.g., real-time editing tools, interactive dashboards, retail checkout)
```
Nit: I think this was my example, but in this context "retail checkout" might be too vague for users to really understand what kind of checkout experience we mean. Maybe we should just cut that one?

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+  - Routes with deep user interactions where constant server updates provide little benefit (e.g., real-time editing tools, interactive dashboards, retail checkout)
+- **Benefits:**
+  - **Reduced server load:** Less demanding on your server resources.
+  - **Better performance for interactive features:** Provides a smoother user experience for dynamic applications.
```
Question: What "interactive features" are more performant in CSR? I would imagine that once the initial page load happens, there's not much difference between a CSR page and an SSR page. Any subsequent interactivity should be fairly equivalent right?

alan-agius4 replied:
True

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+- **Ideal for:**
+  - Routes where data primarily resides on the client-side and is not known to the server (e.g., PWAs with offline functionality)
+  - Routes with deep user interactions where constant server updates provide little benefit (e.g., real-time editing tools, interactive dashboards, retail checkout)
+- **Benefits:**
```
Consider: One potential benefit is not having to make your code SSR-compatible or using SSR-compatible libraries. Just less work for developers.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+  - **Reduced server load:** Content is generated only once, minimizing server resource usage.
+  - **Enhanced cachability:** Static files are easily cached by CDNs, browsers, and even intermediate caching layers for even faster subsequent page loads.
+  - **Scalability:** Deploying SSG files on a CDN improves scalability by offloading traffic from your main application server. This is especially useful for high-traffic websites.
+- **Differs from SSR:**
```
Nit: I think "Differences from SSR" is a little more clear as a heading.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+##### Prerender (SSG)
+- **Ideal for:**
+  - Data primarily resides on the client-side, changes frequently, and doesn't need to be known by the server (e.g., PWAs with offline functionality and local storage).
+  - User interactions are complex and frequent, and constantly updating the server for rendering provides little benefit (e.g., real-time editing tools and interactive dashboards).
```
Consider: This is pretty similar to CSR and might be confusing in that I think you're imagining an SSG render without user data. That's definitely possible, but probably not what most users would imagine, and that use case might be better satisfied by app shell?

I think better use cases might be:
1. Pages which are largely static (ex. a landing page, an about page, blog posts from markdown).
2. Pages with immutable / easily cachable data (changelog entries for a documentation site, some other example I can't think of right now?).

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+  - Data primarily resides on the client-side, changes frequently, and doesn't need to be known by the server (e.g., PWAs with offline functionality and local storage).
+  - User interactions are complex and frequent, and constantly updating the server for rendering provides little benefit (e.g., real-time editing tools and interactive dashboards).
+- **Benefits:**
+  - **Excellent performance:** Provides the fastest loading times, benefiting both user experience and SEO.
```
Nit: I suggest dropping "SEO" from lines where there is another line in the same section talking more directly about SEO, otherwise it feels like we're unnecessarily repeating ourselves / stressing SEO in particular.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+  - **Excellent performance:** Provides the fastest loading times, benefiting both user experience and SEO.
+  - **Improved SEO:** Search engines can easily crawl and index the static HTML pages.
+  - **Reduced server load:** Content is generated only once, minimizing server resource usage.
+  - **Enhanced cachability:** Static files are easily cached by CDNs, browsers, and even intermediate caching layers for even faster subsequent page loads.
```
Suggestion: Spell out "Content Delivery Network (CDN)" here or earlier in the doc.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+  - **Content Generation:** SSR generates content on each request, while SSG generates it once during the build process.
+  - **Cacheability:** SSG offers superior cachability due to its static nature, while SSR's dynamic content is less easily cached.
+  - **Reduced Server Costs:** SSG significantly reduces server load and costs compared to SSR, especially for high-traffic sites.
+
```
Consider: SSG has a potential drawback in slower production builds, since we need to prerender every relevant route. If this grows to 1000+ routes, that can be significant.

(Not sure if it would be worth mentioning that dev builds are not affected, since that's where build times are typically more important, maybe just bold "**production** builds"?)

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
+##### App Shell
+- **Ideal for:**
+  - Providing a fast initial load experience by displaying a basic layout quickly for CSR routes.
+  - Improving perceived performance and reducing the feeling of waiting for the application to load.
```
Consider: Is there a possible benefit to only having to make the shell of the application compatible with SSR/SSG?

alan-agius4 replied:
I don't believe that's entirely accurate. While it's true that the application doesn't need to be fully compatible with SSR and SSG from a runtime perspective, it does need to be from a build perspective. At least not without going out of the way to make it so.

dgp1130 replied:
What would you say are the benefits to app shell over the alternatives? Why would I want to pick that over CSR, SSR, or SSG?

alan-agius4 replied:
By defining a route with `RenderMode.AppShell` a shell is generated for all routes that are marked `RenderMode.CSR`. 

You'd use an AppShell to provide a fast initial load experience by displaying a skeleton layout instead of a blank page.

dgp1130 commented on adev/src/content/guide/hybrid-rendering.md:
````diff
@@ -124,28 +95,63 @@ const serverConfig: ApplicationConfig = {
-```
-
-### Parameterized Routes Prerendering
+### Choosing the right rendering mode
````
Consider: I wonder if this section should actually be the last one (move it to right before "Entry points"). This would allow us to focus on the actual code and features available and keep user attention. Then at the end we can effectively "zoom out" to a broader architectural view of how and when to use each form of rendering.

jelbourn replied:
+1, I generally like pushing recommendations and best practices _after_ more purely explanatory sections

dgp1130 reviewed (approved):
Thanks for expanding the "Choosing the right rendering mode" content, this is looking really helpful. Just a handful of suggestions for making sure we're communicating the right trade offs.

AndrewKushnir commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,325 @@
+})
+export class MyComponent {
+  constructor() {
+    const request = inject(REQUEST, { optional: true }); // Inject with { optional: true }
```
Adding `{ optional: true }` may not be very ergonomic and easy to miss. I'm wondering if we should instead [update the corresponding `InjectionToken`](https://github.com/angular/angular-cli/blob/127811203ac73803c53f7412c0b0880c004f455f/goldens/public-api/angular/ssr/tokens/index.api.md?plain=1#L10) and include `null` into the type (for this and other tokens mentioned above). WDYT?

// cc @alan-agius4 @dgp1130 @jelbourn

alan-agius4 replied:
Agreed, updated

jelbourn commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,325 @@
+### Rendering modes
+
+The Server Routing Configuration allows you to specify how each route in your application should be rendered. The following rendering modes are available:
+- **Server (SSR):** Renders your application on the server for each request, sending a fully populated HTML page to the browser. This is beneficial for SEO and initial load time. See the [Server-Side Rendering (SSR) guide](guide/ssr) for more information.
```
This should be a table

jelbourn commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,325 @@
+### Rendering modes
+
+The Server Routing Configuration allows you to specify how each route in your application should be rendered. The following rendering modes are available:
+- **Server (SSR):** Renders your application on the server for each request, sending a fully populated HTML page to the browser. This is beneficial for SEO and initial load time. See the [Server-Side Rendering (SSR) guide](guide/ssr) for more information.
```
I might keep this section focused on the "what" and defer any "why would I want this" to your later section

E.g., drop the sentence "This is beneficial for SEO and initial load time."

alan-agius4 replied:
Agreed

jelbourn reviewed (commented):
Got part of the way through before I have to head out for the day- I'll come back for the rest later

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
````diff
@@ -0,0 +1,325 @@
+
+This developer preview gives you a first look at these powerful new features. We encourage you to explore them and provide feedback to help shape the future of Angular server rendering.
+
+## Setting up hybrid rendering
+
+To create a **new** project with SSR, run:
+
+```shell
+ng new --ssr --server-routing
+```
+
+To add SSR to an **existing** project, use the Angular CLI `ng add` command.
+
+```shell
+ng add @angular/ssr --server-routing
+```
+
+HELPFUL: A dedicated migration will be provided in the near future to simplify the process of adding hybrid rendering to existing SSR projects. This will offer a more streamlined and robust way to integrate hybrid rendering into your applications.
````
I don't want to place too much emphasis on "server routing." While it's one of the key features, it isn't the only purpose of the new APIs.

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,325 @@
+By evaluating these factors, you can determine the optimal rendering strategy for each route of your application.
+
+##### Server (SSR)
+- **Ideal for:**
```
I think "Ideal" is more appropriate in this case, as it suggests the best match or most suitable option.

AndrewKushnir replied:
I see, thanks for sharing. For me, "Ideal" sets an extremely high expectation and may look prescriptive, so I was trying to soften it a bit, but I don't feel very strongly about this :)

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
````diff
@@ -0,0 +1,325 @@
+export const reqHandler = createNodeRequestHandler(app);
+```
+
+### @angular/ssr/tokens
````
`@angular/ssr` primarily consists of server-side code and logic. By having a separate entry point, we ensure that we don't rely entirely on tree-shaking to exclude code that shouldn't be included in the server bundles.

AndrewKushnir replied:
Should we move those tokens to the `@angular/core` in this case? I feel like the approach of adding extra entrypoints with a very narrow scope might introduce DX challenges in the future.

alan-agius4 replied:
I am not strongly about it. I can do a PR to port them over if it's okay to move them in `@angular/core` or `@angular/platform-browser`. 

Please let me know.

jelbourn commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,324 @@
+- **Server-Side Rendering (SSR):** Redirects are performed using standard HTTP redirects (e.g., 301, 302) within the server-side rendering process.
+- **Prerendering (SSG):** Redirects are implemented as "soft redirects" using `<meta http-equiv="refresh">` tags in the prerendered HTML. This allows for redirects without requiring a round trip to the server.
+
+### Choosing a rendering mode
```
I was trying to see if this section would make sense in a table, and I accidentally ended up expanding it out to paragraph form:

```markdown
### Choosing a rendering mode

Each rendering mode has different benefits and drawbacks. You can choose rendering modes based on
the specific needs of your application.

#### Client-side rendering

Client-side rendering has the simplest development model, as you can write code that assumes
it always runs in a web browser. This lets you use a wide range of client-side libraries that
also assume they run in a browser.

Client-side rendering generally has worse performance than other rendering modes, as it must
download, parse, and execute your page's JavaScript before the user can see any rendered content.
If your page fetches more data from the server as it renders, users also have to wait for those
additional requests before they can view the complete content.

If your page is indexed by search crawlers, client-side rendering may negatively affect search
engine optimization (SEO), as search crawlers have limits to how much JavaScript they execute
when indexing a page.

When client-side rendering, the server does not need to do any work to render a page beyond serving
static JavaScript assets. You may consider this factor if server cost is a concern.

Applications that support installable, offline experiences with [service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) can rely on client-side rendering without needing to communicate with a server.

#### Server-side rendering

Server-side rendering offers faster page loads than client-side rendering. Instead of waiting
for JavaScript to download and run, the server directly renders an HTML document upon
receiving a request from the browser. The user experiences only the latency necessary for the
server to fetch data and render the requested page. This mode also eliminates the need for
additional network requests from the browser, as your code can fetch data during rendering on the
server.

Server-side rendering generally has excellent search engine optimization (SEO), as search crawlers
receive a fully rendered HTML document.

Server-side rendering requires you to author code that does not strictly depend on
browser APIs and limits your selection of JavaScript libraries that assume they run in a browser.

When server-side rendering, your server runs Angular to produce an HTML response for every request.
This additional cost may affect server hosting costs.

#### Build-time prerendering

Prerendering offers faster page loads than both client-side rendering and server-side rendering.
Because prerendering creates HTML documents at _build-time_, the server can directly respond to
requests with the static HTML document without any additional work.

Prerendering requires that all information necessary to render a page is available at _build-time_.
This means that prerendered pages cannot include any data to the specific user loading the page.
This means that prerendering is primarily useful for pages that are the same for all users of
your application.

Because prerendering occurs at build-time, it may add significant time to your production builds. Using `getPrerenderParams` to produce a large number of HTML documents may affect the total file size of your deployments, and thus lead to slower deployments.

It
may also add time to your deployments based on the number of static HTML documents included in
your build output.

Prerendering generally has excellent search engine optimization (SEO), as search crawlers
receive a fully rendered HTML document.

Prerendering requires you to author code that does not strictly depend on
browser APIs and limits your selection of JavaScript libraries that assume they run in a browser.

Prerendering incurs extremely little overhead per server request, as your server responds with
static HTML documents. Static files are also easily cached by Content Delivery Networks (CDNs), browsers, and intermediate caching layers for even faster subsequent page loads. Deploying static HTML files to a CDN improves scalability by offloading work from your application web server, which is impactful for high-traffic applications.
```

* I like that writing this out in paragraph for lets us elaborate more naturally on the _why_ for some of these points. 
* I think I captured all the points you had, but it's possible I missed a bullet
* I ran out of time for the day to also write a similar section for App shell, so I'll leave that one to you

alan-agius4 replied:
I tried organizing the information in a table, but it ended up losing a lot of detail because it had to be kept brief.

While paragraph form allows for a fuller explanation of each point, it makes it harder to quickly refer back compared to using bullet points and easily see the pros and cons of each mode. This less skimmable format may make it difficult for users to quickly find and comprehend the information they need.

jelbourn commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,324 @@
+- **Benefits:**
+  - **Improved perceived performance:** Users see a basic layout quickly, which gives the impression of a faster loading application.
+
+## Entry-points
```
The header "Entry points" doesn't line up with a specific developer intent / action. Is there a different header that maps more directly to what a developer might be looking for or trying to accomplish?

alan-agius4 replied:
Updating all the below handing

- Configuring a Node.js Server
- Configuring a non-Node.js Server
- Accessing Request and Response via DI

jelbourn replied:
These are great 👍

AndrewKushnir commented on adev/src/content/guide/hybrid-rendering.md:
````diff
@@ -0,0 +1,324 @@
+
+```shell
+ng add @angular/ssr --server-routing
+```
````
Quick question: would this command configure server routing for existing apps that use SSR already? If not, can we recommend a way to enable it manually for such apps?

alan-agius4 replied:
No, this will not configure the server routing to existing SSR apps. We'll provide a migration for this in the near future.

AndrewKushnir commented on adev/src/content/guide/ssr.md:
```diff
@@ -14,32 +14,26 @@ The main advantages of SSR as compared to client-side rendering (CSR) are:
 
 To create a **new** project with SSR, run:
```
I think it might be helpful for discoverability if we add a short section about server routing to the SSR guide (a couple sentences and a link to the detailed guide).

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,324 @@
+
+### Configuring server routes
+
+You can create a server route config by declaring an array of [`ServerRoute`](api/ssr/ServerRoute 'API reference') objects. This configuration typically lives in a file that `app.routes.server.ts`.
```
It's always `app`.

alan-agius4 commented on adev/src/content/guide/hybrid-rendering.md:
```diff
@@ -0,0 +1,324 @@
+- **Performance:** How quickly do you need the page to load for users?
+- **Search Engine Optimization (SEO):** How important is search engine visibility for this content?
+- **Server resources:** What level of server resources can you dedicate to rendering?
+- **Interactivity:** How dynamic and interactive is the content on this page?
+- **Content updates:** How frequently will the content on this page change?
```
Removed you are correct, especially with event reply.

jelbourn reviewed (approved):
LGTM on the content ✔️ 

Separate from this PR, I think we might be able to improve the navigation structure of this larger section; we can follow up on slack

alan-agius4 commented:
@jessicajaniuk done

alan-agius4 committed:
fixup! build(docs-infra): include API references for `@angular/ssr` entry points

thePunderWoman commented:
This PR was merged into the repository by commit ade9ba789c7cdb21b1c9e60794aefc817df40a92.

The changes were merged into the following branches: main, 19.0.x
